lib/core/src/Cardano/Wallet/Network.hs

```diff
@@ -361,57 +361,65 @@ follow nl tr cps0 yield header =
+    step delay hasRolledForward cursor = runExceptT (nextBlocks nl cursor) >>= \case
         Left e -> do
             traceWith tr $ MsgNextBlockFailed e
-            sleep (retryDelay delay) cps cursor
+            sleep (retryDelay delay) hasRolledForward cursor
 
         Right AwaitReply -> do
             traceWith tr MsgSynced
-            sleep delay0 cps cursor
+            sleep delay0 hasRolledForward cursor
 
         Right (RollForward cursor' _ []) -> do -- FIXME Make RollForward return NE
             traceWith tr MsgSynced
-            sleep delay0 cps cursor'
+            sleep delay0 hasRolledForward cursor'
 
         Right (RollForward cursor' tip (blockFirst : blocksRest)) -> do
             let blocks = blockFirst :| blocksRest
             traceWith tr $ MsgApplyBlocks (header <$> blocks)
             params <- getProtocolParameters nl
             action <- yield blocks (tip, params)
             traceWith tr $ MsgFollowAction (fmap show action)
-            continueWith cursor' [header $ NE.last blocks] action
+            continueWith cursor' True action
 
         Right (RollBackward cursor') ->
-            -- NOTE
-            -- In case the node asks us to rollback to the last checkpoints we
-            -- gave, we take no action and simply move on to the next query.
-            -- This happens typically with the Haskell nodes which always
-            -- initiates the protocol by asking clients to rollback to the last
-            -- known intersection.
-            case (cursorSlotId nl cursor', cps) of
-                (_, []) ->
-                    step delay0 cps cursor'
-                (sl, _:_) | sl == slotId (last cps) ->
-                    step delay0 cps cursor'
-                (sl, _) ->
-                    destroyCursor nl cursor' $> FollowRollback sl
+            -- After negotiating a tip, the node asks us to rollback to the
+            -- intersection. We may have to rollback to our /current/ tip.
+            --
+            -- This would do nothing, but @follow@ handles rollback by exiting
+            -- such that a new negotiation is required, leading to an infinite
+            -- loop.
+            --
+            -- To handle this we check if we have ever rolled forward in this
+            -- follow-session.
+            --
+            -- NOTE: We could compare the rollback slot to the last cps0, or if
+            -- [] to `SlotId 0 0`. The problem is that the wallet associates
+            -- `SlotId 0 0` with both the first slot and the genesis/origin
+            -- point. (NOTE: I do believe this is correct, but not 100% sure)
+            --
+            -- Some alternative solutions would be to:
+            -- 1. Make sure we have a @BlockHeader@/@SlotId@ for @Origin@
+            -- 2. Stop forcing @follow@ to quit on rollback
+            if hasRolledForward
+            then destroyCursor nl cursor' $> FollowRollback sl
+            else step delay0 cps cursor'
```

Comment:
I like the approach, but the part checking for 

```hs
                (sl, _:_) | sl == slotId (last cps) ->
            --
                    step delay0 cps cursor'
```

is actually necessary. Imagine the following scenario:

- We have a set of checkpoints `[a,b,c]` initially provided to find an intersection. 
- The node find an intersection with `b`, and ask to rollback to `b`. 

At this point, if we don't acknowledge this rollback, on the next attempt to roll forward, the worker will crash because the block yielded won't be a valid continuation. The worker will restart itself, provided `[a,b,c]` as a list of checkpoints... and loop ad infinitum.

Reply:
> is actually necessary.

🤦‍♂️damn! Yeah, you're right.